src/librustc/ty/mod.rs

```diff
@@ -264,42 +265,67 @@ impl AssocItem {
     }
 }
 
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, HashStable)]
+struct AssocItemByNameUnhygienic(ty::AssocItem);
```

Comment:
It's somewhat weird to use a set with a comparator as a map, if we can use an actual map instead.
This should be an implementation detail of `OrderedIndexMultiMap`, IMO.
(Which can reuse all of the implementation of `OrderedIndexMultiSet` because they are pretty much equivalent `Map<Key, T>` ~ `Set<(Key, T)>`).
The extra space for the `Symbol` key shouldn't be a big deal.

Reply:
How will `in_definition_order` return a slice then?

Reply:
Yeah, I meant, why is it necessary to use a slice specifically, as opposed to an opaque `impl Iterator`.